File: lib/facility.ts
import * as z from 'zod';

// 座標
export const degreeType = z.number();
export type Degree = z.infer<typeof degreeType>;

// 日付
export const dateType = z.string();

// 緯度経度
export const coordinateType = z.object({
    // 緯度
    lat: degreeType,

    // 経度
    lon: degreeType,
});
export type Coordinate = z.infer<typeof coordinateType>;

// 御朱印情報
export const goshuinType = z.object({
    // 最低1つ以上の画像
    pictureUrls: z.array(z.string()).min(1),

    // 説明文 (なくても良し)
    description: z.string().nullable().or(z.undefined()),

    // 日付文字列
    date: dateType,
});
export type Goshuin = z.infer<typeof goshuinType>;

// 宗教施設の種別
export const facilityKindType = z.union([
    // 寺
    z.literal('temple'),

    // 神社
    z.literal('shrine'),
]);
export type FacilityKind = z.infer<typeof facilityKindType>;

// 貰ったパンフだとかをくっつける
export const attachmentType = z.object({
    // 画像か何かのURL
    mediaUrl: z.string(),
    
    // 日付
    date: dateType,
});
export type Attachment = z.infer<typeof attachmentType>;

// 宗教施設
export const facilityType = z.object({
    // id なんでも良いので適当にユニークなのを自分でつける
    id: z.string(),

    // 施設の名前
    name: z.string(),

    // 種別
    kind: facilityKindType,

    // 座標
    coordinate: coordinateType,

    // 御朱印のリスト
    goshuinList: z.array(goshuinType),

    // メモ
    memo: z.string().nullable().or(z.undefined()),

    // 付属物
    attachments: z.array(attachmentType).nullable().or(z.undefined()),
});
export type Facility = z.infer<typeof facilityType>;

const idsType = z.array(z.string());

// idをすべて取得する
export const getFacilityIds = async (): Promise<Array<string>> => {
    const { default: ids } = await import('../lib/facilities.json');
    return idsType.parse(ids);
};

// 一件取得する
export const getFacility = async (id: string): Promise<Facility> => {
    const { default: facility } = await import(`../lib/facilities/${id}.json`);
    return facilityType.parse(facility);
};

File: bot/src/facility.rs
use serde::{Serialize, Deserialize};
use chrono::{Date, Utc};

/// 宗教施設の種別
#[derive(Serialize, Deserialize, Debug)]
pub enum FacilityKind {
    /// 寺院
    #[serde(rename = "temple")]
    Temple,

    /// 神社
    #[serde(rename = "shrine")]
    Shrine,
}

/// 座標
#[derive(Serialize, Deserialize, Debug)]
pub struct Coordinate {
    /// 緯度
    pub lat: f64,

    /// 経度
    pub lon: f64,
}

impl Coordinate {
    pub fn new(lat: f64, lon: f64) -> Coordinate {
        Coordinate { lat, lon }
    }
}

/// 御朱印
#[derive(Serialize, Deserialize, Debug)]
pub struct Goshuin {
    /// 画像のURL
    #[serde(rename = "pictureUrls")]
    pub picture_urls: Vec<String>,
    
    /// 説明 メモとか
    pub description: Option<String>,

    /// 日付
    pub date: String,
}

impl Goshuin {
    pub fn new(
        picture_urls: Vec<String>,
        date: Date<Utc>,
        description: Option<String>
    ) -> Goshuin {
        Goshuin {
            description,            
            picture_urls,
            date: date.to_string(),
        }
    }
}

/// 付属物
#[derive(Serialize, Deserialize, Debug)]
pub struct Attachment {
    /// 画像のURL
    #[serde(rename = "mediaUrl")]
    pub media_url: String,

    /// 日付
    pub date: String,
}

impl Attachment {
    pub fn new(
        media_url: String,
        date: Date<Utc>
    ) -> Attachment {
        Attachment {
            media_url,
            date: date.to_string(),
        }
    }
}

/// 宗教施設情報
#[derive(Serialize, Deserialize, Debug)]
pub struct Facility {
    /// id
    pub id: String,

    /// 施設の名前
    pub name: String,

    /// 種別
    pub kind: FacilityKind,

    /// 座標
    pub coordinate: Coordinate,

    /// 御朱印のリスト
    #[serde(rename = "goshuinList")]
    pub goshuin_list: Vec<Goshuin>,

    /// なんかメモとか
    pub memo: Option<String>,

    /// 付属物
    pub attachments: Option<Vec<Attachment>>,
}

impl Facility {
    pub fn new(
        id: String,
        name: String,
        kind: FacilityKind,
        coordinate: Coordinate,
        goshuin_list: Vec<Goshuin>,
        memo: Option<String>,
        attachments: Option<Vec<Attachment>>
    ) -> Facility {
        Facility {
            id,
            name,
            kind,
            memo,
            coordinate,
            attachments,
            goshuin_list,
        }
    }
}

#[test]
fn test_parse() {
    let json = r#"
    {
        "id": "hirose-taisha",
        "name": "廣瀬大社",
        "kind": "shrine",
        "coordinate": {
            "lat": 34.5913983,
            "lon": 135.7483875
        },
        "goshuinList": [
            {
                "pictureUrls": ["/images/IMG_6327.jpg"],
                "description": "",
                "date": "2021-10-24"
            }
        ]
    }
    "#;

    let facility = serde_json::from_str::<Facility>(&json);
    dbg!(&facility);

    assert!(facility.is_ok());
}
